### PR TITLE
8233451: (fs) Files.newInputStream() cannot be used with character special files

### PR DESCRIPTION
**Pull request by bplb**
Add `isOther` and `available` methods to `FileChannelImpl` and the interfaces to native code and use these in `ChannelInputStream` to work around cases where a wrapped `FileChannelImpl` is not really seekable.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8233451](https://bugs.openjdk.org/browse/JDK-8233451): (fs) Files.newInputStream() cannot be used with character special files (**Enhancement** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21508/head:pull/21508` \
`$ git checkout pull/21508`

Update a local copy of the PR: \
`$ git checkout pull/21508` \
`$ git pull https://git.openjdk.org/jdk.git pull/21508/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21508`

View PR using the GUI difftool: \
`$ git pr show -t 21508`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21508.diff">https://git.openjdk.org/jdk/pull/21508.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21508#issuecomment-2412520747)

### CODE REVIEW

File: src/java.base/windows/native/libnio/ch/FileDispatcherImpl.c [hunk 1]
```diff
@@ -392,6 +394,57 @@ Java_sun_nio_ch_FileDispatcherImpl_size0(JNIEnv *env, jobject this, jobject fdo)
     return (jlong)size.QuadPart;
 }
 
+JNIEXPORT jint JNICALL
+Java_sun_nio_ch_FileDispatcherImpl_available0(JNIEnv *env, jobject this, jobject fdo)
+{
+    HANDLE handle = (HANDLE)(handleval(env, fdo));
+    jlong available;
+    if (handleAvailable((jlong)handle, &available)) {
```
Comment (AlanBateman) [inline src/java.base/windows/native/libnio/ch/FileDispatcherImpl.c:402]: Ugh, we shouldn't be using handleAvailable here. This native method needs to stand on its own.

File: src/java.base/windows/native/libjava/io_util_md.h [hunk 2]
```diff
@@ -41,7 +41,7 @@ WCHAR* fileToNTPath(JNIEnv *env, jobject file, jfieldID id);
 JNIEXPORT WCHAR* getPrefixed(const WCHAR* path, int pathlen);
 WCHAR* currentDir(int di);
 int currentDirLength(const WCHAR* path, int pathlen);
-int handleAvailable(FD fd, jlong *pbytes);
+JNIEXPORT int handleAvailable(FD fd, jlong *pbytes);
```
Comment (AlanBateman) [inline src/java.base/windows/native/libjava/io_util_md.h:44]: This should not be exported.

**Comment by AlanBateman** [comment]
Overall the approach is good, except for the Windows implementation of "available" that will need re-work. For ChannelInputStream.isOther then it's effective a stable field, we can do some improvements there.

**Comment by archiecobbs** [comment]
Albeit with my very basic understanding of these internals, it seems wrong that the `FileChannel` created to represent a special ends up also being a `SeekableFileChannel`. This seems like the true root of the problem - that is, the fact that these files are being labeled as seekable but they're not actually seekable is what leads to the bug.

The approach also this leads to the requirement for "fixups" like these, which are brittle and kindof a red flag that something is not quite right:
```java
if (!(ch instanceof SeekableByteChannel sbc) ||
     (ch instanceof FileChannelImpl fci && isOther()))
```

In other words, we already have an interface for representing exactly this situation, so why not use it? What is the rationale for not doing so? I'm not saying there isn't one, just seeking to better understand the trade-offs...

**Comment by AlanBateman** [comment]
> Albeit with my very basic understanding of these internals, it seems wrong that the `FileChannel` created to represent a special ends up also being a `SeekableFileChannel`. This seems like the true root of the problem - that is, the fact that these files are being labeled as seekable but they're not actually seekable is what leads to the bug.

FileChannel implements SeekableByteChannel. You'll see exactly the same thing if you use FileChannel.open to open a special file.

File: src/java.base/windows/native/libjava/io_util_md.h (hunk 2 above)
Comment (bplb) [reply src/java.base/windows/native/libjava/io_util_md.h:44]: Okay.

File: src/java.base/windows/native/libnio/ch/FileDispatcherImpl.c (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/windows/native/libnio/ch/FileDispatcherImpl.c:402]: When/if we get to the point that FIS/FOS/RAF is implemented on FileChannel then it will become clearer.

**Comment by archiecobbs** [comment]
> FileChannel implements SeekableByteChannel. You'll see exactly the same thing if you use FileChannel.open to open a special file.

Ah... thanks. Also (for example) `FileInputStream.getChannel()` returns a `FileChannel`, so we're stuck with it.

OK so this is an "original sin" situation which we can't fix here. Thanks for the clarification.

**Comment by xxDark** [comment]
Hello.
I see that my [report](https://bugs.openjdk.org/browse/JDK-8341666) was closed as a duplicate of JDK-8233451 (the bug this PR fixes).
This does not resolve the situation with `FileInputStream`?
`java.io` will not be fixed? Practically none of the new API such as `readNBytes` could be used as it relies on calls to `lseek` on unix systems (this affects MacOS as well).
Thanks.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java [hunk 3]
```diff
@@ -53,13 +53,26 @@ class ChannelInputStream extends InputStream {
     private byte[] bs;       // Invoker's previous array
     private byte[] b1;
 
+    private Boolean isOther = null;
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:56]: I don't particularly like using a Boolean for tri-states but it's not too bad here. No need to initialize to null. It could be Stable but probably not much benefit here all usages require file I/O that dominates.

Are you going to add a comment to this field as readers might know now what "other" means? In the APIs we say "something other than a regular file, directory, or symbolic link" and maybe that could be useful here.

File: src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java [hunk 4]
```diff
@@ -529,6 +530,46 @@ public long size() throws IOException {
         }
     }
 
+    public int available() throws IOException {
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java:533]: This can be package-private. It would be useful to add a method description as FC doesn't define this method, same thing for isOther.

File: src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java [hunk 5]
```diff
@@ -529,6 +530,46 @@ public long size() throws IOException {
...
+        try {
+            beginBlocking();
+            ti = threads.add();
+            if (!isOpen())
+                return false;
+            boolean attempted = Blocker.begin();
+            try {
+                return isOther = nd.isOther(fd);
+            } finally {
+                Blocker.end(attempted);
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java:565]: No need for Blocker begin/end here, we only use for direct I/O or file locking.

File: src/java.base/unix/native/libnio/ch/UnixFileDispatcherImpl.c [hunk 6]
```diff
@@ -178,6 +181,57 @@ Java_sun_nio_ch_UnixFileDispatcherImpl_size0(JNIEnv *env, jobject this, jobject
...
+    struct stat fbuf;
+
+    if (fstat(fd, &fbuf) == -1)
+        handle(env, -1, "isOther failed");
+
+    if (S_ISREG(fbuf.st_mode) || S_ISDIR(fbuf.st_mode) || S_ISLNK(fbuf.st_mode))
+        return JNI_FALSE;
+
+    return JNI_TRUE;
+}
```
Comment (AlanBateman) [inline src/java.base/unix/native/libnio/ch/UnixFileDispatcherImpl.c:233]: Can you check these syscalls for EINTR? Might be okay, but worth checking.

File: src/java.base/windows/native/libnio/ch/FileDispatcherImpl.c [hunk 7]
```diff
@@ -392,6 +394,74 @@ Java_sun_nio_ch_FileDispatcherImpl_size0(JNIEnv *env, jobject this, jobject fdo)
...
+}
+
+
+JNIEXPORT jboolean JNICALL
+Java_sun_nio_ch_FileDispatcherImpl_isOther0(JNIEnv *env, jobject this, jobject fdo)
+{
+    HANDLE handle = (HANDLE)(handleval(env, fdo));
+
+    BY_HANDLE_FILE_INFORMATION finfo;
+    GetFileInformationByHandle(handle, &finfo);
```
Comment (AlanBateman) [inline src/java.base/windows/native/libnio/ch/FileDispatcherImpl.c:438]: This returns a BOOL that needs to be checked.

File: test/jdk/java/nio/file/Files/InputStreamTest.java [hunk 8]
```diff
@@ -123,6 +128,17 @@ static void testSkip(Path tmpdir) throws IOException {
...
 
+    /**
+     * Tests that Files.newInputStream(Path).available() does not throw
+     */
+    static void testAvailable() throws IOException {
+        Path stdin = Path.of("/dev", "stdin");
+        if (Files.exists(stdin)) {
+            InputStream s = Files.newInputStream(stdin);
+            s.available();
+        }
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/InputStreamTest.java:139]: I assume you meant to close the input stream.

What about the other methods defined by InputStream that have special handling in ChannelInputStream? I assume we should add test coverage for these methods.

**Comment by bplb** [comment]
> Either the problem reported in [JDK-8341666](https://bugs.openjdk.org/browse/JDK-8341666) will be addressed in the context of this PR, or that issue will be reopened and addressed by a subsequent PR.

@xxDark I reopened [JDK-8341666](https://bugs.openjdk.org/browse/JDK-8341666) which will be addressed by a separate PR. Thanks for pointing out our oversight.

File: src/java.base/unix/native/libnio/ch/UnixFileDispatcherImpl.c (hunk 6 above)
Comment (bplb) [reply src/java.base/unix/native/libnio/ch/UnixFileDispatcherImpl.c:233]: None of them are documented to so fail in the Linux and macOS man pages.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java (hunk 3 above)
Comment (bplb) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: Fixed in a163d8d and 492659f.

File: src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java (hunk 4 above)
Comment (bplb) [reply src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java]: Fixed in a163d8d.

File: test/jdk/java/nio/file/Files/InputStreamTest.java (hunk 8 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/InputStreamTest.java:149]: I added test coverage for `skip` but not for `readAllBytes` nor `readNBytes`:
- `readAllBytes` throws before but hangs after the change;
- `readNBytes` hangs both before and after the change.

File: src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java (hunk 5 above)
Comment (bplb) [reply src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java]: Fixed in a163d8d.

File: test/jdk/java/nio/file/Files/InputStreamTest.java (hunk 8 above)
Comment (AlanBateman) [reply test/jdk/java/nio/file/Files/InputStreamTest.java:149]: I wonder if there is another special file that could be read to EOF to allow the methods be tested. Asking because the overrides have a code path we need to test.

**Review by AlanBateman — APPROVED** [review]
The latest update looks okay, maybe we might find a better special file so the readAllXXX methods can be tested.

File: test/jdk/java/nio/file/Files/InputStreamTest.java (hunk 8 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/InputStreamTest.java:149]: I need to investigate this. I might hold off integrating even thought it's approved until this is determined.

Comment (bplb) [reply test/jdk/java/nio/file/Files/InputStreamTest.java:149]: > I wonder if there is another special file that could be read to EOF to allow the methods be tested.

I think that this can done using a named pipe, at least on UNIX.

File: test/jdk/java/nio/file/Files/InputStreamTest.java [hunk 9]
```diff
@@ -22,37 +22,126 @@
  */
 
 /* @test
- * @bug 8227609
+ * @bug 8227609 8233451
  * @summary Test of InputStream and OutputStream created by java.nio.file.Files
- * @library ..
+ * @library .. /test/lib
+ * @build jdk.test.lib.Platform
+ * @run junit InputStreamTest
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/Files/InputStreamTest.java:29]: I think this will need --enable-native-access=ALL-UNNAMED.

**Review by AlanBateman — APPROVED** [review]
This update looks good, just a bit minor nits. I assume you'll run the test many tests on all platforms to ensure its stability before integrating.

File: test/jdk/java/nio/file/Files/InputStreamTest.java (hunk 9 above)
Comment (bplb) [reply test/jdk/java/nio/file/Files/InputStreamTest.java:29]: It works as is in the CI.

File: test/jdk/java/nio/file/Files/InputStreamTest.java [hunk 10]
```diff
@@ -22,37 +22,126 @@
...
+        }
+    }
+
+    private static Thread createWriteThread() {
+        Thread t = new Thread(
+            new Runnable() {
+                public void run() {
+                    try {
+                        try (FileOutputStream fos = new FileOutputStream(PIPE);)
+                        {
```
[the remark this answers is not in the document]
Comment (bplb) [reply test/jdk/java/nio/file/Files/InputStreamTest.java:99]: Thanks, I'll clean it up.

**Comment by bplb** [comment]
> I assume you'll run the test many tests on all platforms to ensure its stability before integrating.

I already have to an extend but will run again with more platforms and repeats.

**Comment by lbergelson** [comment]
👍 Thanks for these changes!

**Comment by bplb** [comment]
> I assume you'll run the test many tests on all platforms to ensure its stability before integrating.

I ran the version of commit f4ab451 30 times on all platforms without failures.

**Review by AlanBateman — APPROVED** [review]

**Comment by bplb** [comment]
30 repeats on all platforms succeeded also after the most recent commit, 2876d9f.
